src/Core/Managed/Shared/TelemetryClient.cs

```diff
@@ -371,6 +373,11 @@ public void Initialize(ITelemetry telemetry)
             if (string.IsNullOrEmpty(instrumentationKey))
             {
                 instrumentationKey = this.configuration.InstrumentationKey;
+
+                if (string.IsNullOrEmpty(instrumentationKey))
+                {
+                    instrumentationKey = PlatformSingleton.Current.GetEnvironmentVariable(InstrumentationKeyWebSitesEnvironmentVariable);
```

Comment:
I agree on caching, discussing the priority ordering, but I don't think it should change during runtime.